src/components/RegistrationForm/RegistrationForm.tsx

```diff
@@ -0,0 +1,221 @@
+
+import group_of_people from '../../assets/images/group_of_people.png';
+import { FormInput } from '../FormInput/FormInput';
+import style from './RegistrationForm.module.scss';
+
+const INITIAL_FORM_DATA = {
+  firstName: '',
+  lastName: '',
+  email: '',
+  password: '',
+  confirmPassword: '',
+};
+
+const INITIAL_ERRORS = {
+  firstName: false,
+  lastName: false,
+  email: false,
+  password: false,
+  confirmPassword: false,
+};
+
+type FormData = typeof INITIAL_FORM_DATA;
+type FormErrors = typeof INITIAL_ERRORS;
+
+const ERROR_MESSAGES = {
+  firstName: 'First name cannot be empty or contain numbers',
+  lastName: 'Last name cannot be empty or contain numbers',
+  email: 'Please enter a valid email',
+  password:
+    'Password must have at least 8 characters, 1 uppercase letter and 1 number',
+  passwordMatch: "Passwords don't match",
+};
+
+const validators = {
+  name: (value: string): boolean => {
+    if (value.length === 0) {
+      return false;
+    }
+    for (let i = 0; i < value.length; i++) {
+      if (value[i] >= '0' && value[i] <= '9') {
+        return false;
+      }
+    }
+
+    return true;
+  },
+  email: (value: string): boolean => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value),
+  password: (value: string): boolean => {
+    let hasUpperCase = false;
+    let hasNumber = false;
+    if (value.length < 8) {
+      return false;
+    }
+    for (let i = 0; i < value.length; i++) {
+      if (value[i] >= '0' && value[i] <= '9') {
+        hasNumber = true;
+      }
+      if (value[i] >= 'A' && value[i] <= 'Z') {
+        hasUpperCase = true;
+      }
+    }
+
+    return hasUpperCase && hasNumber;
+  },
+};
+
+export const RegistrationForm = () => {
+  const [formData, setFormData] = useState<FormData>(INITIAL_FORM_DATA);
+  const [shakeAnimation, setShakeAnimation] = useState(false);
+  const [isSubmitted, setIsSubmitted] = useState(false);
+  const [errors, setErrors] = useState<FormErrors>(INITIAL_ERRORS);
+
+  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    setFormData((prev) => ({ ...prev, [name]: value }));
+
+    const fieldValidation = {
+      firstName: () => !validators.name(value),
+      lastName: () => !validators.name(value),
+      email: () => !validators.email(value),
+      password: () => ({
+        password: !validators.password(value),
+        confirmPassword: formData.confirmPassword !== value,
+      }),
+      confirmPassword: () => ({
+        confirmPassword: value !== formData.password,
+      }),
+    };
+
+    const fieldValidationResult =
+      fieldValidation[name as keyof typeof fieldValidation]();
+
+    setErrors((prev) => ({
+      ...prev,
+      ...(typeof fieldValidationResult === 'boolean'
+        ? { [name]: fieldValidationResult }
+        : fieldValidationResult),
+    }));
+  };
+
+  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    const newErrors = {
```

Comment:
If the above returns just booleans and these are unified, we could move this whole object into function above the component. Combine it with the validators. I will show you

Reply:
The handleChange and handleSubmit need different validation logic - during typing we don't want to validate confirmPassword, but during submit we need to check if passwords match. That's why I prefer to keep these separate.